### ISSUE TITLE
 Exception while decompiling e000:0000:  AddressOutOfBoundsException

### ISSUE THREAD

**Issue by Godzil** [NationalSecurityAgency/ghidra#239]
When  trying to disassemble a function I get that error message in the Decompile window:

  Exception while decompiling e000:0000: ghidra.program.model.address.AddressOutOfBoundsException:
       Offset must be between 0x0 and 0x10ffef, got 0x250236 instead!

The code is 80(1)86 code, so 16 real mode, I don't clearly understand what cause that error, and especially how to look to know what cause that error. There is no JMP in the function code that would or should cause such an issue, and the error does not say anything on where this happen, and I can't find any logs that shows more information.

All I can get for now is the full Java stack frame for that exception:

```
2019-03-21	13:55:07	ERROR	(DecompileProcess) Unexpected Exception: Offset must be between 0x0 and 0x10ffef, got 0x250236 instead! ghidra.program.model.address.AddressOutOfBoundsException: Offset must be between 0x0 and 0x10ffef, got 0x250236 instead!
			at ghidra.program.model.address.AbstractAddressSpace.makeValidOffset(AbstractAddressSpace.java:717)
			at ghidra.program.model.address.GenericAddressSpace.makeValidOffset(GenericAddressSpace.java:21)
			at ghidra.program.model.address.GenericAddress.<init>(GenericAddress.java:55)
			at ghidra.program.model.address.SegmentedAddress.<init>(SegmentedAddress.java:72)
			at ghidra.program.model.address.SegmentedAddressSpace.getAddress(SegmentedAddressSpace.java:289)
			at ghidra.program.model.address.SegmentedAddressSpace.getAddress(SegmentedAddressSpace.java:25)
			at ghidra.program.model.pcode.Varnode.readXMLAddress(Varnode.java:662)
			at ghidra.app.decompiler.DecompileCallback.getMappedSymbolsXML(DecompileCallback.java:605)
			at ghidra.app.decompiler.DecompileProcess.getMappedSymbolsXML(DecompileProcess.java:700)
			at ghidra.app.decompiler.DecompileProcess.readResponse(DecompileProcess.java:315)
			at ghidra.app.decompiler.DecompileProcess.sendCommand1ParamTimeout(DecompileProcess.java:530)
			at ghidra.app.decompiler.DecompInterface.decompileFunction(DecompInterface.java:701)
			at ghidra.app.decompiler.component.Decompiler.decompile(Decompiler.java:57)
			at ghidra.app.decompiler.component.DecompilerManager.decompile(DecompilerManager.java:167)
			at ghidra.app.decompiler.component.DecompileRunnable.monitoredRun(DecompileRunnable.java:108)
			at ghidra.util.task.RunManager$RunnerJob.doExecute(RunManager.java:334)
			at ghidra.util.task.RunManager$RunnerJob.run(RunManager.java:309)
			at ghidra.util.worker.AbstractWorker$JobCallback.process(AbstractWorker.java:133)
			at ghidra.util.worker.AbstractWorker$JobCallback.process(AbstractWorker.java:123)
			at generic.concurrent.ConcurrentQ$CallbackCallable.call(ConcurrentQ.java:655)
			at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
			at generic.concurrent.FutureTaskMonitor.run(FutureTaskMonitor.java:70)
			at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
			at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
			at java.base/java.lang.Thread.run(Thread.java:834)
```

Where should I look for to try to identify what cause that issue?

Thanks

**Comment by emteere** [comment]
The decompiler is attempting to come back and get information for a particular address.  You may be able to see what the address being requested and why using Debug Function Decompilation.  The output file is all the information used in the decompilation.  Most likely this won't be useful to you, but if you can post the .XML file produced it could help us to debug the situation.

Review the information before you post, as it will have label names, type names, and the actual bytes.

You also post the byte string copied from your example and provide the address of the bytes and the processor in use.  It may be the address is a large negative number, or there is something strange with the segment/offset in the address that can't map into a segmented address.

**Comment by Godzil** [comment]
Here is the XML file, 
[megStart.zip](https://github.com/NationalSecurityAgency/ghidra/files/2996277/megStart.zip)

**Comment by emteere** [comment]
@Godzil if you get a chance, can you see if the 9.1 branch or master has fixed your issue.  There has been a bit of re-work to the 16-bit code handling for x86 and also to support HCS12 paging.
If you aren't building, then we may be putting out a beta soon if internal testing goes well, or if not that then the 9.1 release.

**Comment by Godzil** [comment]
Sorry I cannot build ghidra at the moment.

**Comment by davidgiven** [comment]
I am seeing precisely this on Ghidra 9.1. In my case, I get:

`ghidra.program.model.address.AddressOutOfBoundsException: Offset must be between 0x0 and 0x10ffef, got 0x198400c6 instead!`

In my case, the variable being looked at is at address 00C6, and I've used the register manager to enforce DS=1984, so the segmented address of the variable is at 1984:00C6... which matches the address in the error. So I think what's happening is that something is failing to handle segmented addresses properly.

For me this happens when I attempt to follow the symbol to its definition. It doesn't matter whether I do this from the decompiler or from the disassembler, I get the same error.

I see that makeValidOffset() is no longer at line 717 of AbstractAddressSpace.java. Does this mean this has been fixed? If so, do you know where I can get an interim set of binaries to try?

**Comment by davidgiven** [comment]
Never mind --- I figured out how to build it (thank you for the excellent installation instructions and smooth build! Although I did have to patch gradlew into the repository first --- Gradle 5's not in Debian yet. And, weirdly, none of the text shown in the UI has spaces). Yes, this manifests in head of git. The stack trace is attached below.

```
Uncaught Exception! 
AddressOutOfBoundsException - Offset must be between 0x0 and 0x10ffef, got 0x198425ea instead!
ghidra.program.model.address.AddressOutOfBoundsException: Offset must be between 0x0 and 0x10ffef, got 0x198425ea instead!
	at ghidra.program.model.address.AbstractAddressSpace.makeValidOffset(AbstractAddressSpace.java:623)
	at ghidra.program.model.address.GenericAddressSpace.makeValidOffset(GenericAddressSpace.java:21)
	at ghidra.program.model.address.GenericAddress.<init>(GenericAddress.java:55)
	at ghidra.program.model.address.SegmentedAddress.<init>(SegmentedAddress.java:62)
	at ghidra.program.model.address.SegmentedAddressSpace.getAddress(SegmentedAddressSpace.java:285)
	at ghidra.program.model.address.SegmentedAddress.getNewAddress(SegmentedAddress.java:111)
	at ghidra.app.plugin.core.decompile.actions.RenameVariableAction.getStorageAddress(RenameVariableAction.java:77)
	at ghidra.app.plugin.core.decompile.actions.RenameVariableAction.isEnabledForDecompilerContext(RenameVariableAction.java:165)
	at ghidra.app.plugin.core.decompile.actions.AbstractDecompilerAction.lambda$isEnabledForContext$0(AbstractDecompilerAction.java:50)
	at ghidra.app.plugin.core.decompile.DecompilerActionContext.checkActionEnablement(DecompilerActionContext.java:105)
	at ghidra.app.plugin.core.decompile.actions.AbstractDecompilerAction.isEnabledForContext(AbstractDecompilerAction.java:49)
	at docking.ComponentPlaceholder.contextChanged(ComponentPlaceholder.java:482)
	at docking.DockingWindowManager.contextChanged(DockingWindowManager.java:2110)
	at docking.AbstractDockingTool.contextChanged(AbstractDockingTool.java:209)
	at ghidra.app.plugin.core.decompile.DecompilerProvider.contextChanged(DecompilerProvider.java:626)
	at ghidra.app.plugin.core.decompile.DecompilerProvider.locationChanged(DecompilerProvider.java:486)
	at ghidra.app.decompiler.component.DecompilerController.locationChanged(DecompilerController.java:245)
	at ghidra.app.decompiler.component.DecompilerPanel.fieldLocationChanged(DecompilerPanel.java:674)
	at docking.widgets.fieldpanel.FieldPanel$CursorHandler.notifyCursorChanged(FieldPanel.java:2105)
	at docking.widgets.fieldpanel.FieldPanel$CursorHandler.setCursorPos(FieldPanel.java:1899)
	at docking.widgets.fieldpanel.FieldPanel$FieldPanelMouseAdapter.mousePressed(FieldPanel.java:1297)
	at java.desktop/java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:287)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6629)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6397)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5008)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4544)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

---------------------------------------------------
Build Date: 2019-Dec-08 0003 CET
Ghidra Version: 9.2
Java Home: /usr/lib/jvm/java-11-openjdk-amd64
JVM Version: Debian 11.0.5
OS: Linux 5.3.0-2-amd64 amd64
Workstation: hilfy.lan

```

**Comment by davidgiven** [comment]
I did a bit of debugging. The actual disassembly looks like this:

```
mov ax, 0xc6
push ax
call something
```

`something` takes a data pointer as a parameter, but ghidra doesn't know that the 0xc6 is a segment offset until `something` gets seen and the types unified. At this point, to turn that 0xc6 into an address the current value of DS needs to be looked up, producing 0x1984, to produce 1984:00c6.

I think what's happening is that whatever is responsible for doing this is just gluing the value of DS on the front producing 0x198400c6, which is an invalid segment offset, rather than producing a proper segmented address. But I have no idea where that might be so I don't think I'm going to get any further.

Let me know if there's anything you need...

**Comment by Godzil** [comment]
Interestingly, I've just downloaded 9.1 and load the same project. 

Now the function that was not decompiling produce something, but some of the label/pointer are incorrect, and when clicked they produce a popup with the same error as before:

```
Uncaught Exception! 
AddressOutOfBoundsException - Offset must be between 0x0 and 0x10ffef, got 0xffffffffec252000 instead!
ghidra.program.model.address.AddressOutOfBoundsException: Offset must be between 0x0 and 0x10ffef, got 0xffffffffec252000 instead!
	at ghidra.program.model.address.AbstractAddressSpace.makeValidOffset(AbstractAddressSpace.java:717)
	at ghidra.program.model.address.GenericAddressSpace.makeValidOffset(GenericAddressSpace.java:21)
	at ghidra.program.model.address.GenericAddress.<init>(GenericAddress.java:55)
	at ghidra.program.model.address.SegmentedAddress.<init>(SegmentedAddress.java:50)
	at ghidra.program.model.address.SegmentedAddressSpace.getAddressInSegment(SegmentedAddressSpace.java:131)
	at ghidra.program.model.address.SegmentedAddress.getNewAddress(SegmentedAddress.java:109)
	at ghidra.app.plugin.core.decompile.actions.RenameVariableAction.getStorageAddress(RenameVariableAction.java:77)
	at ghidra.app.plugin.core.decompile.actions.RenameVariableAction.isEnabledForDecompilerContext(RenameVariableAction.java:165)
	at ghidra.app.plugin.core.decompile.actions.AbstractDecompilerAction.lambda$isEnabledForContext$0(AbstractDecompilerAction.java:50)
	at ghidra.app.plugin.core.decompile.DecompilerActionContext.checkActionEnablement(DecompilerActionContext.java:105)
	at ghidra.app.plugin.core.decompile.actions.AbstractDecompilerAction.isEnabledForContext(AbstractDecompilerAction.java:49)
	at docking.ComponentPlaceholder.contextChanged(ComponentPlaceholder.java:482)
	at docking.DockingWindowManager.contextChanged(DockingWindowManager.java:2110)
	at docking.AbstractDockingTool.contextChanged(AbstractDockingTool.java:209)
	at ghidra.app.plugin.core.decompile.DecompilerProvider.contextChanged(DecompilerProvider.java:626)
	at ghidra.app.plugin.core.decompile.DecompilerProvider.locationChanged(DecompilerProvider.java:486)
	at ghidra.app.decompiler.component.DecompilerController.locationChanged(DecompilerController.java:245)
	at ghidra.app.decompiler.component.DecompilerPanel.fieldLocationChanged(DecompilerPanel.java:674)
	at docking.widgets.fieldpanel.FieldPanel$CursorHandler.notifyCursorChanged(FieldPanel.java:2105)
	at docking.widgets.fieldpanel.FieldPanel$CursorHandler.setCursorPos(FieldPanel.java:1899)
	at docking.widgets.fieldpanel.FieldPanel$FieldPanelMouseAdapter.mousePressed(FieldPanel.java:1297)
	at java.desktop/java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:287)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6629)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6397)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5008)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4544)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

---------------------------------------------------
Build Date: 2019-Dec-18 1306 EST
Ghidra Version: 9.1.1
Java Home: /usr/lib/jvm/java-11-openjdk-amd64
JVM Version: Oracle Corporation 11.0.3
OS: Linux 4.19.0-4-amd64 amd64
Workstation: 
```
I can't get exactly what is causing that . It also happen something while clicking on the assembly window.

Here is an exemple of what I think cause the issue:

```asm
       e000:00f9 44              INC        SP
       e000:00fa 44              INC        SP
       e000:00fb 57              PUSH       DI
       e000:00fc 6a 00           PUSH       0x0
       e000:00fe ff 76 e8        PUSH       word ptr [BP + display_segment]
       e000:0101 ff 76 e6        PUSH       word ptr [BP + display_offset]
       e000:0104 e8 b3 aa        CALL       _memset                                          Clear display
```

The decompiled code look like that:

```C
_memset((void *)&UNK_e000_e250,display_segment,0x20000000);
```

Clicking on the UNK_e000_e250 will trigger the popups.

_memset is using that prototype:

```C
void * __cdecl16near _memset(void *__s,int __c,size_t __n);
```
I'm not sure where does the E250h offset come from.



Edit: Ok found an explanation there.
The _memset function in fact should be 

```C
void * __cdecl16near _memset(void far*__s,int __c,size_t __n);
```

But apart from using "pointer32" that don't really make any sense as it is not a 32bit linear pointer, the decompiler was getting confused with the stack content.

It seems there is still no way to explicitely say this is a "far pointer" or this is a "near pointer". I don't think the `pointer32` and `pointer16` are well suited for that, especially because they don't give the real type behing (`pointer32` does not say it is a `char far *` for example)

**Comment by spheenik** [comment]
I have also stumbled upon this error, and started to dig into it. I can reproduce it with the git version from yesterday, and want to contribute some findings. 

I also disassemble a 16bit DOS real mode program. My decompiler input looks like this

https://pastebin.com/2ucbLkm0

The decompiler output is too big for pastebin, but the problem seems to be that some address varnodes are incorrectly generated. I find the following in the decompile output:

```xml
<addr space="const" offset="0x3f2b0002" size="4" ref="0x1bf6"/>
<addr space="const" offset="0x3f2b683c" size="4" ref="0x1bfb"/>
<addr space="const" offset="0x3f2b688e" size="4" ref="0x1c00"/>
<addr space="const" offset="0x3f2b689d" size="4" ref="0x1c06"/>
<addr space="const" offset="0x3f2b689f" size="4" ref="0x1c0b"/>
<addr space="const" offset="0x3f2b68af" size="4" ref="0x1c10"/>
<addr space="const" offset="0x3f2b68af" size="4" ref="0x1c15"/>
<addr space="const" offset="0x3f2b68d0" size="4" ref="0x1c1a"/>
<addr space="const" offset="0x3f2b6cf4" size="4" ref="0x1c1f"/>
<addr space="const" offset="0x3f2b6cf4" size="4" ref="0x1c24"/>
<addr space="const" offset="0x3f2b6cf6" size="4" ref="0x1c29"/>
<addr space="const" offset="0x3f2b6df4" size="4" ref="0x1c2e"/>
```

The data segment in this case is 0x3f2b, so for the segmented memory model, the first address for example should be 0x3f2b2 (0x3f2b << 4 + 0x0002) 

If I patch the code to fix this after the fact here:

https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/address/AbstractAddressSpace.java#L625-L626

where instead of throwing, I do the following:

```java
		long fixed = ((offset & 0xFFFF0000L) >>> 12) + (offset & 0x0000FFFFL);
		return fixed;
```
the decompilation succeeds.


**FWIW:**

The failing PcodeOpAST seem to always be of the following format:

`(unique, 0x10000616, 4) PTRSUB (const, 0x0, 4) , (const, 0x3f2b2c0e, 4)`

where the second parameter contains a GenericAddress (and I would think it should be a SegmentedAddress?)

**Comment by sad-dev** [comment]
It appears that the decompiler itself is producing invalid addresses, the above 'patch' being a workaround,

**Comment by ryanmkurtz** [comment]
Fixed by 0e81327c46adbbe12c4d78c470624fc077922124
